fplll/pruner/pruner.h

```diff
@@ -0,0 +1,709 @@
+      }
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      if (target < 0.0)
+      {
+        target      = 0.99;
+        opt_overall = true;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
+    // normalization on |b_i^*|
+    load_basis_shape(gso_r);
+  }
+
+  /**
+     @brief Constructor for Pruner with multiple basis (averaging the cost function over all of
+     them)
+
+     Algorithmic details are given for the wrapping function prune().
+
+     @param enumeration_radius radius of the desired enumeration
+     @param preproc_cost cost of preprocessing (i.e. additive cost for a retrying an enumeration)
+     @param gso_rs vector of vector Gram-Schmidt length (squared) of the enumerated block over
+     several bases
+     @param target desired target success probability/expected solutions after all retrial.
+     @param metric metric is to be optimized : PRUNER_METRIC_PROBABILITY_OF_SHORTEST or
+     PRUNER_METRIC_EXPECTED_SOLUTIONS
+     @param flags complementary parameters : PRUNER_CVP  PRUNER_START_FROM_INPUT  PRUNER_GRADIENT
+     PRUNER_NELDER_MEAD  PRUNER_VERBOSE
+  */
+
+  Pruner(const FT enumeration_radius, const FT preproc_cost, const vector<vector<double>> &gso_rs,
+         const FT input_target     = 0.9,
+         const PrunerMetric metric = PRUNER_METRIC_PROBABILITY_OF_SHORTEST,
+         int flags                 = PRUNER_GRADIENT)
+      : enumeration_radius(enumeration_radius), preproc_cost(preproc_cost), target(input_target),
+        metric(metric), flags(flags)
+  {
+    verbosity = flags & PRUNER_VERBOSE;
+    n         = gso_rs[0].size();
+    d         = n / 2;
+    if (flags & PRUNER_CVP)
+    {
+      symmetry_factor = 1;
+    }
+    min_pruning_coefficients.resize(d);
+    btmp.resize(d);
+    bftmp.resize(n);
+    fill(min_pruning_coefficients.begin(), min_pruning_coefficients.end(), 0.);
+    set_tabulated_consts();
+    // need to fix target if possible
+    if (metric == PRUNER_METRIC_PROBABILITY_OF_SHORTEST)
+    {
+      // if input_target > 1 or < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (target > 1.0 || target < 0.0)
+      {
+        target      = 0.99;
+        opt_overall = true;
+      }
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      // if input_target < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (target < 0.0)
+      {
+        target      = 0.99;
+        opt_overall = true;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
+    load_basis_shapes(gso_rs);
+  }
+
+  /**
+     Run the optimization process, successively using the algorithm activated
+
+     @brief run the optimization process
+  */
+  void optimize_coefficients_evec(/*io*/ vector<double> &pr);
```

Comment:
It would be nice to understand the hierarchy and differences between all these functions.

Reply:
Yep, updated.